src/triangular.jl

```diff
@@ -559,3 +562,129 @@ end
+                push!(code.args, :($(X[i,j]) = $z))
+            else
+                ex = :(A.data[$(LinearIndices(sa)[i,i])] * B.data[$(LinearIndices(sb)[i,j])])
+                for k = i+1:j
+                    ex = :($ex + A.data[$(LinearIndices(sa)[i,k])] * B.data[$(LinearIndices(sb)[k,j])])
+                end
+                push!(code.args, :($(X[i,j]) = $ex))
+            end
+        end
+    end
+
+    return quote
+        @_inline_meta
+        @inbounds $code
+        return UpperTriangular(similar_type(B.data, $TAB)(tuple($(X...))))
+    end
+
+end
+
+@generated function _A_mul_B(::Size{sa}, ::Size{sb}, A::LowerTriangular{<:TA,<:StaticMatrix}, B::LowerTriangular{<:TB,<:StaticMatrix}) where {sa,sb,TA,TB}
+    n = sa[1]
+    if n != sb[1]
+        throw(DimensionMismatch("left and right-hand must have same sizes, got $(n) and $(sb[1])"))
+    end
+
+    X = [Symbol("X_$(i)_$(j)") for i = 1:n, j = 1:n]
+
+    TAB = promote_op(*, eltype(TA), eltype(TB))
+    z = zero(TAB)
+
+    code = quote end
+    for j = 1:n
+        for i = 1:n
+            if i < j
+                push!(code.args, :($(X[i,j]) = $z))
+            else
+                ex = :(A.data[$(LinearIndices(sa)[i,j])] * B.data[$(LinearIndices(sb)[j,j])])
+                for k = j+1:i
+                    ex = :($ex + A.data[$(LinearIndices(sa)[i,k])] * B.data[$(LinearIndices(sb)[k,j])])
+                end
+                push!(code.args, :($(X[i,j]) = $ex))
+            end
+        end
+    end
+
+    return quote
+        @_inline_meta
+        @inbounds $code
+        return LowerTriangular(similar_type(B.data, $TAB)(tuple($(X...))))
+    end
+
+end
+
+
+@generated function _A_mul_B(::Size{sa}, ::Size{sb}, A::UpperTriangular{<:TA,<:StaticMatrix}, B::LowerTriangular{<:TB,<:StaticMatrix}) where {sa,sb,TA,TB}
+    n = sa[1]
+    if n != sb[1]
+        throw(DimensionMismatch("left and right-hand must have same sizes, got $(n) and $(sb[1])"))
+    end
+
+    X = [Symbol("X_$(i)_$(j)") for i = 1:n, j = 1:n]
+
+    code = quote end
+    for j = 1:n
+        for i = 1:n
+            k1 = max(i,j)
+            ex = :(A.data[$(LinearIndices(sa)[i,k1])] * B.data[$(LinearIndices(sb)[k1,j])])
+            for k = k1+1:n
+                ex = :($ex + A.data[$(LinearIndices(sa)[i,k])] * B.data[$(LinearIndices(sb)[k,j])])
+            end
+            push!(code.args, :($(X[i,j]) = $ex))
+        end
+    end
+
+    return quote
+        @_inline_meta
+        @inbounds $code
+        TAB = promote_op(*, eltype(TA), eltype(TB))
+        return similar_type(B.data, TAB)(tuple($(X...)))
+    end
+
+end
+
+@generated function _A_mul_B(::Size{sa}, ::Size{sb}, A::LowerTriangular{<:TA,<:StaticMatrix}, B::UpperTriangular{<:TB,<:StaticMatrix}) where {sa,sb,TA,TB}
+    n = sa[1]
+    if n != sb[1]
+        throw(DimensionMismatch("left and right-hand must have same sizes, got $(n) and $(sb[1])"))
+    end
+
+    X = [Symbol("X_$(i)_$(j)") for i = 1:n, j = 1:n]
+
+    code = quote end
```

Comment:
Do you want a `LineNumberNode`? If not you could just do `code = Expr(:block)` instead.

Reply:
I was just blindly copying the code from the rest of this file; I don't know enough to know whether I should have a preference one way or the other.

Reply:
OK... let's change them all to `Expr(:block)`, then. You'll get a relevant line number from the final quote anyway.

Reply:
I can do that but won't get to it until later today; feel free to make the change yourself in the mean time if you'd rather.